### DISCUSSION TITLE
mosdns 的性能

### DISCUSSION THREAD

**Discussion by IrineSistiana** [IrineSistiana/mosdns#300] [General]
## 匹配器性能

除了正则和 keyword 域名规则，导入再多规则也不会性能影响。匹配一百万条规则用时并不是匹配一条的一百万倍。 大概只有20倍。反正都是几十纳秒(0.000000001s)的事，忽略不计。

## 转发性能

测试程序是随手写的，测试结果看看就好，**不要**作为选择用什么协议什么软件的依据。毕竟 DNS 转发除非应用规模和流量特别特别特别特别特别特别大，否则几乎不可能碰到性能瓶颈。用什么没有区别。

- 单核的 i7-7700HQ (@3.3Ghz) 的虚拟机。测试时间大约是 2021/03，同机回环 IO，没有延时。所有软件的配置都是最简的，仅转发。mosdns 使用 `forward` 和 `fast_forward` 作转发。
- `/` 表示不支持这个协议。
- `?` 表示未测试。
- 单位每秒转发数 `op/s`。

|     | `forward` | `fast_forward` | dnsproxy | smartdns | overture | dnsmasq |
| --- | --------- | -------------- | -------- | -------- | -------- | ------- |
| UDP | 6300      | 24000          | 4900     | 60000   | 7000     | 39000   |
| TCP | 8200      | 27000          | 3800     | 48000   | 8600     | ?       |
| DoT | 21000     | 25000          | 11000    | 56000   | ?        | /       |
| DoH | 14000     | 14000          | 6500     | 24000   | ?        | /       |

---

如果有条件欢迎用 dnsperf 地址: https://github.com/DNS-OARC/dnsperf 等测试工具对 mosdns 进行实测。

**Comment by liangxyz** [comment]
传统的协议相对来说数据稍差，安全协议性能前列，不过家用基本用不到那么大的开销吧？
另外请问作者，我遇到了大量in-addr.arpa报错 context deadline exceeded，请问是咋回事啊？具体可见我新发的https://github.com/IrineSistiana/mosdns/discussions/299

**Comment by InspoOnU** [comment]
嗯，分流对性几乎没有影响。在OpenWRT，RK3399上测试了MosDNS在加载7万条域名规则的情况下，100 op/s时占用CPU约15%。使用DoQ上游，丢包1%。

**Reply by IrineSistiana** [reply]
没有影响。(除非用了正则规则。既然都用正则了，这方面知识基本上都应该很清楚。)

匹配一条域名/IP 和匹配一百万条消耗是差不多的。匹配一百万条规则用时**并不是匹配一条的一百万倍。** 大概只有20倍。反正都是几十纳秒的事，忽略不计。

**Comment by pmkol** [comment]
为什么DOT会比UDP转发数还要多，SSL多一次握手，为什么速度反而更快呢

**Reply by IrineSistiana** [reply]
连接复用不需要握手。
dnsproxy/foward udp 性能差是因为它不会复用套接字。

**Reply by pmkol** [reply]
> 连接复用不需要握手。 dnsproxy/foward udp 性能差是因为它不会复用套接字。

也就是说，上游最快的连接方式，应该是tcp，其次是tls？
enable_pipeline 我看到是支持RFC 7766的
我是mosdns自带的工具测试，发现返回的是可能不支持RFC 7766，但我阿里DNS官网查询，写的是DOT支持RFC 7858，应该兼容RFC 7766吧？

**Reply by IrineSistiana** [reply]
只要没碰到性能瓶颈就不存在所谓的“谁快谁慢”
只要处理器有aes加速，现在基本上都有，tls和tcp看不出区别。
pipeline是7766的建议，非强制标准。

**Reply by pmkol** [reply]
> pipeline

也就是说，阿里DNS可以开启pipeline？即使你的程序检测没通过说可能不支持

**Reply by IrineSistiana** [reply]
不可以。wiki写的很明白
